### DISCUSSION TITLE
Interacting with community

### DISCUSSION THREAD

**Discussion by LecrisUT** [swig/swig#2718] [General]
Hi there @ojwb @wsfulton,

There are so many issues and PR that have been accepted and worked upon, it is a pity that they just fall out fue to radio silence from the project owners. I don't know if it's an issue eith the notifications or something that make these issues get lost in the gaps.

What would help is to have at least 1-2 public meetings per month where people from the community can participate and get some attention or feedback. I've seen many other projects do this, but I think for the current state of this project in particular it would be beneficial.

Most of us younger people are not familiar with mail-lists, so having other means of communication would be helpful

**Comment by ojwb** [comment]
I'm going to convert this to a discussion, since it seems to fit better there - this is not a bug report of a bug in SWIG's code (I'd suggest the development mailing list would really be the best place - there's not much you need to know about mailing lists - you just send an email to the list address and it gets sent on to a list of people).

The key problem (at least for me, but I suspect this also applies to William) is simply lack of time to work on SWIG.  SWIG is not my main FOSS project and nobody is paying for my time to deal with SWIG issues or patched.  I'm involved with SWIG because I use it, but rather than only dealing with the parts I actually use I do what I can more generally, typically fitted into short gaps while I wait for something to run.

How will allocating some of my already very limited SWIG time for these public meetings help?  It's not going to create more time for me to work on SWIG, so it seems it'll inevitably mean I actually get less done on SWIG, plus it's not much fun to have to tell people once or twice a month that I'm sorry but I still haven't had time to look at their issue or patch.  If you're thinking of real-time meetings, timezones are likely to be challenging too as there's no overlap between my working day and Europe's.

The seemingly ever increasing pile of open SWIG issues is not great, but I don't think there's a magical solution.  We could close issues as "stale" after a certain time period of inattention, but I know how demotivating that is to bug reporters from being on the other side of that with other projects.

I do sometimes close unfixed issues manually, but only when there's a good reason - e.g. not enough info to reproduce and the reporter hasn't responded.  Perhaps we could try something like requiring a self-contained reproducer and closing issues without one, as attempting to reproduce an issue without one can take a significant amount of time.

**Reply by LecrisUT** [reply]
This is not about the issues but about the PRs. I want to help out to get better documentation with sphinx (I know how to solve the dead links problem with linkcheck), the github actions (that one is done, just need approval on the approach), and cmake integration (this is the main thing I want to help with). I have tried to ping you there for attention a few times.

I guess the main issue is where people keep their eyes on, and currently that seems to be only the mailing list and new github issues. It would be helpful to have some other means of communication. Something like projects would be nice where someone (I can volunteer some time initially) can triage issues and assign (max 2-3) to a senior developer when things are ready to be reviewed.

Also mailing lists are not ideal, I've recently tried to send a mail on a fedora mailing list that got silently dropped, even though the Cc went through. Some other useful means of communication are on Matrix (can still have IRC bridge) and Discord.

**Comment by LecrisUT** [comment]
Another related aspect is the pushing directly to the main branch, without triggering and waiting for the CI tests. If we use PRs even for 1 commit, that can help a lot in the future (creating changelog from the PRs, avoiding conflicts and duplications, etc.) The actions can take a long time, but with the refactoring in #2647, the test time will be significantly less and you can just set to automerge when all non-experimental jobs pass.

**Comment by vadz** [comment]
Sorry, not sure if it's going to be useful, but I'd like to leave my 0.02€.

Personally, I can understand lack of time (oh yes), but I have trouble determining the priorities of SWIG project. What I've seen time and time again were important, even critical, PRs for mainstream languages (Python, C#, Java, ...) languishing for years without any explanation while not insignificant time was spent on backends for Octave and such (no disrespect to Octave developers intended, obviously, but this is objectively important to much smaller number of people). I'm not going to tell anybody how should they spend their time, of course, but I just don't understand this. And I would have definitely appreciated knowing more about the priorities of the project before spending time on PRs that never got anywhere.

The worst thing is that it used to be not like this and I did manage to get several big PRs integrated. But it seems to be getting worse and worse, or at least it seemed to until I just stopped trying to get anything into SWIG a couple of years ago (to the big relief of everybody, I suspect). And at this stage, the only realistic solutions I see are to

1. Give more people permission to merge PRs into SWIG.
2. Fork SWIG under a more active/available maintainer.

To be perfectly candid, I considered doing (2) myself several times in the past but I didn't want to do it as long as there was still any chance of the current project picking up speed as it would have been disloyal and also potentially unproductive. However I just don't see what else could be done.

**Comment by wsfulton** [comment]
Yes, @ojwb summarises well.  There just isn't enough time to deal with all the issues and pull requests.  Overall, priority is given to keeping the build system working, that is, making sure we at least stand still and keep new versions of target languages working.  I go through phases of prioritising pull requests over development work.  But usually I prioritise a pull request if it includes a testcase or reproducer/motivating example and documentation and feel like I can contribute something useful. Unfortunately a lot of pull requests are inadequate, perhaps we should just close them saying so instead of leaving them languishing?  I also cannot comment on many merge requests without spending quite a bit of effort on them, so I have to mentally filter some out, sorry.  I also tend not to tackle pull requests that are big as they can suck up days if not weeks of my available contribution time at the expense of other work.  Most pull requests are fairly small though, but the large ones are hard. When time is limited, someone will of course be disappointed, that's the reality of it, we just are not physically able to keep everyone happy. I and many many others have put in thousands of hours of effort to SWIG, I'd hope that that is appreciated.

With regard to additional maintainers, I'd love to find them and welcome them. Anyone who comes along and contributes a number of quality pull requests are good candidates, but they come and go very quickly or don't exist. So for anyone looking to move things along, please demonstrate this and put yourself forward. @ojwb has done a great job recently keeping things ticking over.

@LecrisUT, if you are specifically concerned with your pull requests, I'll comment on them there.

A tip for those who have an old pull request and want to bump it up and you feel it is complete and want to get some attention, add it to the next release milestone. I will definitely look at it and comment on it before the release.

We currently have 6% of pull requests open, that's doesn't seem too awful given the complexity of SWIG and the lack of resources.
@vadz, you're a top contributor and it looks like you have 4% of yours waiting. They can be addressed by both of us. There was a time when I seemed to be doing nothing but attending to yours! 

If anyone has concrete suggestions to improve within the bounds we have, then please raise a discussion. @LecrisUT, it's good you have raised the suggestion of public meetings, but in this case, I concur with @ojwb. Suggestions for online community prioritisation and explanations of how to prioritise are welcomed as certainly, the critisms around radio silence and prioritisations that are not transparent are valid. Hopefully my first paragraph in this response helps a bit in the understanding.

**Reply by LecrisUT** [reply]
> A tip for those who have an old pull request and want to bump it up and you feel it is complete and want to get some attention, add it to the next release milestone.

The issue is that the community does not have write access to the milestone, so there has to be a contributor that can triage these issues. If the project can get some help triaging the issues and not push directly to main branch, that would be helpful for the community to not feel ignored. Something very simple like [this](https://github.com/orgs/spglib/projects/1) would be a good first step.

>  Overall, priority is given to keeping the build system working, that is, making sure we at least stand still and keep new versions of target languages working.

Well this is what I am trying to address with #2647 to move more of the maintenance to the already established Github Actions, among many other things. The main thing is that there are such PRs that can greatly minimize the maintenance and complexity, but how are we as external contributors able to point these things out when we do not have a means of reaching out to the main developers, or even anybody who has any write privilege.

As for my PRs in particular, they are at a stage where the questions are very specific to the maintainers:
- What workflow/languages are considered experimental?
- What minimum C/C++ standard should it support, both for building swig itself and for generating the bindings? What is the rolling/deprecation plan for these?
- At this point converting the documentation to sphinx is more feasible, and for the [blocking issue](https://github.com/swig/swig/pull/1844#issuecomment-1704878763) there is a way to detect and correct these issues with `linkcheck`. If we can pause the documentation for 1-2 weeks, I can do all of the conversion and check manually, but there has to be a clear plan that things can be merged in a timely manner, otherwise redoing this work manually will be a lot of effort.
- The CI, and many build processes can be greatly simplified with more migration to CMake. Could we discuss this aspect?

**Comment by vadz** [comment]
Having PRs passing CI is a given, this is not a problem.

IMO the problem is that some PRs (_not_ all of them, but it seems to be impossible to predict for which of them this will happen) are held up until they achieve some impossibly high standards, which *never* happens and they just die off. I am not one to embrace "worse is better" philosophy usually, but I do think that in SWIG case it would make a lot of good to lean a bit more in this direction.

If you want a practical solution, I could definitely volunteer to merge some PRs and be responsible for fixing problems due to them later (on the "best effort" basis, I do have a bus factor of only 1, after all). This would apply to my own PRs, of course, but not only, e.g. I'd be glad to help SWIG to transition to any other docs format because absolutely anything would be better than writing HTML by hand like cavemen. But I can't afford to invest time into this unless I can be sure that this work is merged at the end provided some clear, objective criteria are met.
